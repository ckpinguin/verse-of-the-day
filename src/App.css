.App {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 10px;
  /* justify-content: center; */
  /* grid-template-rows: 100px 200px 100px;   */
  grid-template-rows: minmax(min-content, auto);
  /* used to define rows or columns the implicit grid (if needed) */
  /* grid-auto-rows: minmax(min-content, auto); */
  /* grid-auto-rows: 0.5fr 1fr 0.5fr; */
  
  grid-template-areas:
  ".....      title      .....    "
  "content    content    content  "
  ".....      handle     .....    "
  "footer     footer     footer   ";
}

@media screen and (max-width: 40em) {
    .App {
        grid-template-areas:
        "title      title      .....    "
        "content    content    content  "
        "handle     handle     handle   "
        "footer     footer     footer   ";
    }
}

.App-Title {
    grid-area: title;
/*   grid-column: 1 / 4;
  grid-row: 1; */
}

.App-Main {
    grid-area: content;
/*   grid-column: 1 / 4;
  grid-row: 2; */
}

.App-DatePicker {
    grid-area: handle;
/*   grid-column: 2;
  grid-row: 3; */
}

.App-Footer {
  grid-area: footer;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}
