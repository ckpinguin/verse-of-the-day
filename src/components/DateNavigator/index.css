.Navigator {
    display: grid;
    /* align-items: center; */
    grid-template-columns: 0.5fr 2fr 0.5fr;
    /* grid-gap: 20px; */
    /* justify-content: center; */
    /* grid-template-rows: 100px 200px 100px;   */
    /* grid-template-rows: minmax(min-content, auto); */
    /* used to define rows or columns the implicit grid (if needed) */
    /* grid-auto-rows: minmax(min-content, auto); */
    /* grid-auto-rows: 0.5fr 1fr 0.5fr; */
    
}
