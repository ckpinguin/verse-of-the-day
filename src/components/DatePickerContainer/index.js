import DatePickerContainer from './DatePickerContainer';

export default DatePickerContainer;