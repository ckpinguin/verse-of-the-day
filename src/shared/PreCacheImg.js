import PropTypes from 'prop-types';

PreCacheImg.propTypes = {
    images: PropTypes.array.isRequired,
};

function PreCacheImg(props) {
    const precache = (images) => {
        let image;
        images.map((img) => { // eslint-disable-line array-callback-return
            image = new Image(); // eslint-disable-line no-undef
            image.src = img;
        });
        return false;
    };

    return precache(props.images);
}

export default PreCacheImg;