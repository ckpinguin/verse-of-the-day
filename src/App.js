import React, { Component } from 'react';
import './App.css';

import { isDebug, debug } from './debug';

import './App.css';

import Show from './components/Show/Show';
// import ToggleDisplay from './components/ToggleDisplay/ToggleDisplay';
import Title from './components/Title/Title';
import DateHelper from './shared/DateHelper';
import DatePickerContainer from './components/DatePickerContainer/DatePickerContainer';
import ImageView from './components/ImageView/ImageView';

export default class App extends Component {
    // API-Key for bibleserver.com API with https://our-daily-bread.herokuapp.com/
    static bibleServerAPIKey = 'ad9b36e57ef40242a89170d6f54322c68d6ed86b';
    constructor(props) {
        super(props);
        this.title = 'Our daily bread';
        this.urlBase = 'https://logos.com/media/VerseOfTheDay/768x432/';
        this.state = {
            date: props.date ? props.date : DateHelper.randomDate()
        };
    }

    render() {
        const formattedDate = DateHelper.formatDate(this.state.date, '-');
        const url = this.urlBase + formattedDate + '.png';
        return (  
            <div className="App" style={ isDebug ? debug.borderStyle : {} } >
                <div className="App-Title">
                    <Show if={true}>
                        <Title
                            value={this.title}
                        />
                    </Show>
                </div>
                <div className="App-Main">
                    <Show if={true}>
                        <ImageView
                            url={url}
                        />
                    </Show>
                </div>
                <div className="App-DatePicker">
                    <DatePickerContainer
                        date={this.state.date}
                        onChange={(newDate) => this.updateDate(newDate)}
                    />
                </div>
            </div>
        );
    }

    updateDate(newDate) {
        console.log('Updating state with: ', newDate);
        this.setState ({
            date: newDate
        });
        console.log('state is now: ', this.state);
    }

    componentDidMount() {
    }
    
}
