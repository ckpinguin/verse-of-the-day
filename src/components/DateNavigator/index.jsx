import React from 'react';
import PropTypes from 'prop-types';

import { isDebug, debug } from '../../debug';

import './index.css';

import DateHelper from '../../shared/DateHelper';
import DatePickerContainer from '../DatePickerContainer';
import NavButton from '../NavButton';

DateNavigator.propTypes = {
    date:           PropTypes.object.isRequired,
    onChangeDate:   PropTypes.func.isRequired,
};
function DateNavigator({ date, onChangeDate }) {
    const handleDayPlus = () => onChangeDate(DateHelper.getDateWithChangedDays(date, +1));const handleDayMinus = () => onChangeDate(DateHelper.getDateWithChangedDays(date, -1));
    return (
        <div
            className="DateNavigator"
            style={isDebug ? debug.borderStyle : {}}
        >
            <NavButton onClick={handleDayMinus} icon="fa fa-chevron-left" />
            <DatePickerContainer 
                date={date}
                onChangeDate={onChangeDate}
            />
            <NavButton onClick={handleDayPlus} icon="fa fa-chevron-right" />
        </div>
    );
}
export default DateNavigator;