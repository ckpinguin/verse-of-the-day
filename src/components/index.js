import Title from './Title';
import ImageView from './ImageView';
import DateNavigator from './DateNavigator';
import Show from './Show';
import ToggleDisplay from './ToggleDisplay';
import DatePickerContainer from './DatePickerContainer';
import NumberChooser from './NumberChooser';
import NavButton from './NavButton';

export {
    DateNavigator,
    DatePickerContainer,
    ImageView,
    NavButton,
    NumberChooser,
    Show,
    Title,
    ToggleDisplay
};