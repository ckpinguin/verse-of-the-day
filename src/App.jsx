import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { HotKeys } from 'react-hotkeys';

import { isDebug, debug } from './debug';

import './App.css';

import PreCacheImg from './shared/PreCacheImg';
import { randomDate, formatDate, getDateWithChangedDays } from './shared/DateHelper';
import Show from './components/Show';
import { Dummy } from './components/ToggleDisplay';
import Title from './components/Title';
import DateNavigator from './components/DateNavigator';
import ImageView from './components/ImageView';


export default class App extends Component {
    // API-Key for bibleserver.com API with https://our-daily-bread.herokuapp.com/
    static bibleServerAPIKey = 'ad9b36e57ef40242a89170d6f54322c68d6ed86b';

    static propTypes = {
        date:   PropTypes.object
    }
    static defaultProps = {
        date:   randomDate()
    }
    
    constructor(props) {
        super(props);
        this.title = 'Our daily bread';
        this.urlBase = 'https://logos.com/media/VerseOfTheDay/768x432/';
        this.state = {
            date: props.date 
        };
        // Handlers:
        this.updateDate = this.updateDate.bind(this);
    }

    updateDate(newDate) {
        console.log('App: updating state: ', newDate);
        // Don't forget, setState() is async!
        this.setState ({
            date: newDate
        });
    }
  
    getImgUrl() {
        const formattedDate = formatDate(this.state.date, '-');
        const url = this.urlBase + formattedDate + '.png';
        const urlNext = this.urlBase + formatDate(this.state.date, '-');
        return url;
    }

    render() {
        // TODO: Make the standard up/down key binding work again in subcomponent NumberChooser
        const keyMap = {
            'dayPlus':     ['up', 'right'],
            'dayMinus':   ['down', 'left']
        };
        const handlers = {
            'dayPlus':     () => this.updateDate(getDateWithChangedDays(this.state.date, +1)),
            'dayMinus':   () => this.updateDate(getDateWithChangedDays(this.state.date, -1))
        };
        const url = this.getImgUrl();
        return (  
            <HotKeys
                keyMap={keyMap}
                handlers={handlers}
                className="App"
                style={ isDebug ? debug.borderStyle : {} }
            >
                <div className="App-Title">
                    <Show if>
                        <Title
                            value={this.title}
                        />                       
                    </Show>
                </div>
                <div className="App-Main">
                    <PreCacheImg
                        images={[
                            url,
                        ]}
                    /> 
                    <Show if>
                        <ImageView url={url} />
                    </Show>
                </div>
                <div className="App-DatePicker">
                    <DateNavigator
                        date={this.state.date}
                        onChangeDate={this.updateDate}
                    />
                </div>
            </HotKeys>
        );
    }
}