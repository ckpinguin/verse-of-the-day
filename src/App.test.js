import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme'; 
import renderer from 'react-test-renderer';

import App from './App';
import { ImageView, DateNavigator, Title } from './components';
import PreCacheImg from './shared';

// Mock 3rd-party-library (used as HOC)
jest.mock('react-hotkeys', () => {
    return {
        HotKeys: 'hotkeys'
    };
});

describe('App', () => {
    let props;
    let mountedApp;
    let shallowApp;
    const app = () => {
        if (!mountedApp) {
            mountedApp = mount(
                <App {...props} />
            );
        }
        return mountedApp;
    };
    const appShallow = () => {
        if (!shallowApp) {
            shallowApp = shallow(
                <App {...props} />
            );
        }
        return shallowApp;
    };
    
    beforeEach(() => {
        props = {
            date: new Date(2017, 5, 12)
        };
        mountedApp = undefined;
        shallowApp = undefined;
    });

    it('calls componentDidMount (just for example)', () => {
        App.prototype.componentDidMount = jest.fn();
        app();
        // const instance = wrapper.instance();
        expect(App.prototype.componentDidMount).toHaveBeenCalled();
    });

    it('renders shallow', () => {
        appShallow();
        // expect(wrapper.contains(welcome)).toEqual(true);
        // simpler version for this:
        // expect(wrapper).toContainReact(welcome);
    });

    it('renders without crashing', () => {
        ReactTestUtils.renderIntoDocument(<App {...props} />);
    });    

    it('mounts successfully (full DOM rendering)', () => {
        app();
    });

    it('always renders 4 top-level divs', () => {
        const divs = appShallow().find('div'); // shallow goes only 1 level deep
        expect(divs.length).toBe(4);
    });

    it('always renders `.App-Title` div', () => {
        const div = app().find('div .App-Title');
        expect(div.length).toBe(1);
    });
    describe('the rendered `.App-Title` div', () => {
        const div = () => app().find('div .App-Title');

        it('contains everything else that gets rendered', () => {  
            // const div = app().find('div .App-Title'); // This hangs!
            const div = appShallow().find('div .App-Title');            
            
            // When using .find, enzyme arranges the nodes in order such
            // that the outermost node is first in the list. So we can
            // use .first() to get the outermost div.
            const wrappingDiv = div.find('div .App-Title').first();
            // console.log(divs.first().debug());
            // console.log(appShallow().childAt(1).debug());
            // Enzyme omits the outermost node when using the .children()
            // method on app(). This is annoying, but we can use it
            // to verify that wrappingDiv contains everything else this
            // component renders.
            expect(wrappingDiv).toEqual(appShallow().childAt(0));
        });
        it('always renders a `Title`', () => {
            expect(div().find(Title).length).toBe(1);
        });
        it('`Title` receives `title` prop', () => {
            const title = (div().find(Title));
            expect(Object.keys(title.props()).length).toBe(1);
            expect(title.props().value).toBe('Our daily bread');
        });
    });

    it('always renders `.App-Main div`', () => {
        const div = app().find('div .App-Main');
        expect(div.length).toBe(1);
    });
    describe('the rendered `.App-Main` div', () => {
        const div = () => app().find('div .App-Main');
        
        it('contains everything else that gets rendered', () => {
            const div = appShallow().find('div .App-Main');
            const wrappingDiv = div.first();
            expect(wrappingDiv).toEqual(appShallow().childAt(1));
        });
        it('always renders a `ImageView`', () => {
            expect(div().find(ImageView).length).toBe(1);
        });
    });

    it('always renders `.App-Handle` div', () => {
        const div = app().find('div .App-Handle');
        expect(div.length).toBe(1);
    });
    describe('the rendered `.App-Handle` div', () => {
        const div = () => app().find('div .App-Handle');

        it('contains everything else that gets rendered', () => {
            const div = appShallow().find('div .App-Handle');
            const wrappingDiv = div.first();
            expect(wrappingDiv).toEqual(appShallow().childAt(2));
        });
        it('always renders a `DateNavigator`', () => {
            expect(div().find(DateNavigator).length).toBe(1);
        });
        it('`DateNavigator` receives 2 props', () => {
            const dateNav = (div().find(DateNavigator));
            expect(Object.keys(dateNav.props()).length).toBe(2);
        });
        it('`DateNavigator` receives `date` prop', () => {
            const dateNav = (div().find(DateNavigator));
            expect(dateNav.props().date).toBe(props.date);
        });
        it('`DateNavigator` receives `onChangeDate` prop', () => {
            const dateNav = (div().find(DateNavigator));
            expect(dateNav.props().onChangeDate).toBeDefined;
            expect(dateNav.props().onChangeDate).toBe(App.propTypes.updateDate);
            
        });
    });

    it('always renders `.App-Footer` div', () => {
        const divs = app().find('div .App-Footer');
        expect(divs.length).toBe(1);
    });
    describe('the rendered `.App-Footer` div', () => {
        const div = () => app().find('div .App-Footer');
        
        it('contains everything else that gets rendered', () => {
            const div = appShallow().find('div .App-Footer');
            const wrappingDiv = div.first();            
            expect(wrappingDiv).toEqual(appShallow().childAt(3));
        });
    });

    it('did not change UI compared to recorded snapshot', () => {
        const component = renderer.create(<App {...props} />);
        // component.state = { date: new Date(2017, 7, 1) };
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

});
