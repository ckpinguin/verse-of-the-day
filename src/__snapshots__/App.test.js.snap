// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> did not change UI compared to recorded snapshot 1`] = `
<div
  className="App"
>
  <div
    className="App-header"
  >
    <span
      style={Object {}}
    >
      <img
        alt="Daily Bread"
        className="App-logo"
        src="https://logos.com/media/VerseOfTheDay/768x432/2017-08-01.png"
      />
    </span>
    <div
      style={Object {}}
    >
      <h2>
        Our daily bread
      </h2>
    </div>
  </div>
  <div>
    <p>
      This is DatePicker stub with date: 
      1.8.2017
    </p>
  </div>
</div>
`;
